### DISCUSSION TITLE
Question: Read a single message using System.IO.Pipelines

### DISCUSSION THREAD

**Discussion by kiminuo** [dotnet/runtime#52180] [Q&A] [answered]
Hi .NET Team,

I'm puzzled a bit byt the sample provided here https://docs.microsoft.com/en-us/dotnet/standard/io/pipelines#read-a-single-message. I wonder about this part:

```csharp
// There's no more data to be processed.
if (result.IsCompleted)
{
    if (buffer.Length > 0)
    {
        // The message is incomplete and there's no more data to process.
        throw new InvalidDataException("Incomplete message.");
    }

    break;
}
```

Specifically, I'm curious when the `break` is hit. I was unable to simulate it in my unit test. Can anybody explain that case to me in layman's terms? :)

Thank you!

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by Tornhoof** [comment, marked as answer]
Write only an empty buffer and complete the pipe:
something like:
```csharp
pipeWriter.Write(ReadOnlySpan<byte>.Empty);
pipeWriter.Complete();
```
The ``TryParseMessage`` method will return false (not a complete message) and the pipe is Completed and the buffer.Length is 0, so no exception, that should hit the break
